Mathlib/RingTheory/MvPowerSeries/PiTopology.lean

```diff
@@ -0,0 +1,243 @@
+@[fun_prop]
+theorem continuous_coeff [Semiring R] (d : σ →₀ ℕ) : Continuous (MvPowerSeries.coeff R d) :=
+  continuous_pi_iff.mp continuous_id d
+
+variable (R) in
+/-- constant_coeff is continuous -/
+theorem continuous_constantCoeff [Semiring R] : Continuous (constantCoeff σ R) :=
+  continuous_coeff R 0
+
+/-- A family of power series converges iff it converges coefficientwise -/
+theorem tendsto_iff_coeff_tendsto [Semiring R] {ι : Type*}
+    (f : ι → MvPowerSeries σ R) (u : Filter ι) (g : MvPowerSeries σ R) :
+    Filter.Tendsto f u (nhds g) ↔
+    ∀ d : σ →₀ ℕ, Filter.Tendsto (fun i => coeff R d (f i)) u (nhds (coeff R d g)) := by
+  rw [nhds_pi, Filter.tendsto_pi]
+  exact forall_congr' (fun d => Iff.rfl)
+
+variable (σ R)
+
+/-- The semiring topology on MvPowerSeries of a topological semiring -/
+@[scoped instance]
+theorem instTopologicalSemiring [Semiring R] [TopologicalSemiring R] :
+    TopologicalSemiring (MvPowerSeries σ R) where
+    continuous_add := continuous_pi fun d => Continuous.comp continuous_add
+      (Continuous.prod_mk (Continuous.fst' (continuous_coeff R d))
+        (Continuous.snd' (continuous_coeff R d)))
+    continuous_mul := continuous_pi fun _ => continuous_finset_sum _ (fun i _ => Continuous.comp
+      continuous_mul (Continuous.prod_mk (Continuous.fst' (continuous_coeff R i.fst))
+        (Continuous.snd' (continuous_coeff R i.snd))))
+
+/-- The ring topology on MvPowerSeries of a topological ring -/
+@[scoped instance]
+theorem instTopologicalRing (R : Type*) [TopologicalSpace R] [Ring R] [TopologicalRing R] :
+    TopologicalRing (MvPowerSeries σ R) :=
+  { instTopologicalSemiring σ R with
+    continuous_neg := continuous_pi fun d ↦ Continuous.comp continuous_neg
+      (continuous_coeff R d) }
+
+variable {σ R}
+
+@[fun_prop]
+theorem continuous_C [Ring R] [TopologicalRing R] :
+    Continuous (C σ R) := by
+  classical
+  apply continuous_of_continuousAt_zero
+  rw [continuousAt_pi]
+  intro d
+  change ContinuousAt (fun y => coeff R d ((C σ R) y)) 0
+  by_cases hd : d = 0
+  · convert continuousAt_id
+    rw [hd, coeff_zero_C, id_eq]
+  · convert continuousAt_const
+    rw [coeff_C, if_neg hd]
+
+theorem variables_tendsto_zero [Semiring R] :
+    Filter.Tendsto (X · : σ → MvPowerSeries σ R) Filter.cofinite (nhds 0) := by
+  classical
+  rw [tendsto_pi_nhds]
+  simp_rw [apply_eq_coeff]
+  intro d s hs
+  replace hs : 0 ∈ s := by simpa only [map_zero] using mem_of_mem_nhds hs
+  simp_rw [Filter.mem_map, Filter.mem_cofinite, ← Set.preimage_compl, coeff_X]
+  by_cases h : ∃ i, d = Finsupp.single i 1
+  · obtain ⟨i, rfl⟩ := h
+    apply (Set.finite_singleton i).subset
+    intro x
+    by_cases hx : x = i <;>
+      simp [Finsupp.single_eq_single_iff, Ne.symm, hx, hs]
+  · convert Set.finite_empty
+    rw [Set.eq_empty_iff_forall_not_mem]
+    intro x
+    rwa [Set.mem_preimage, Set.not_mem_compl_iff, if_neg]
+    tauto
+
+theorem tendsto_pow_zero_of_constantCoeff_nilpotent [CommSemiring R]
```

Comment:
Can you show the stronger statement that if `(constantCoeff σ R f) ^ n` tends to 0, i.e. it is _topologically_ nilpotent, then the same is true of `f`?